src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AQNBuilder.Mono.cs

```diff
@@ -0,0 +1,483 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+/*
+	This TypeNameBuilder is modified from CoreCLR's original.
+	It replaces the C++ bits of the implementation with a faithful C# port.
+	Original files:
+		src/coreclr/src/System.Private.CoreLib/src/System/Reflection/Emit/AQNBuilder.cs
+		src/coreclr/src/vm/typestring.cpp
+		src/coreclr/src/vm/typestring.h
+*/
+
+using System.Collections.Generic;
+using System.Collections;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace System.Reflection.Emit
+{
+	// TypeNameBuilder is NOT thread safe NOR reliable
+	internal class TypeNameBuilder
+	{
+		internal enum Format
+		{
+			ToString,
+			FullName,
+			AssemblyQualifiedName,
+		}
+
+		private static class TypeString
+		{
+			public static bool  ContainsReservedChar (string name)
+			{
+				foreach (char c in name.AsSpan ())
+				{
+					if (c == '\0')
+						break;
+					if (TypeString.IsTypeNameReservedChar (c))
+						return true;
+				}
+				return false;
+			}
+
+			public static bool IsTypeNameReservedChar(char ch)
+			{
+				switch (ch)
+				{
+				case ',':
+				case '[':
+				case ']':
+				case '&':
+				case '*':
+				case '+':
+				case '\\':
+					return true;
+
+				default:
+					return false;
+				}
+			}
+		}
+
+		private class TypeNameBuilderImpl : IDisposable
+		{
+			private enum ParseState
+			{
+				START         = 0x0001,
+				NAME          = 0x0004,
+				GENARGS       = 0x0008,
+				PTRARR        = 0x0010,
+				BYREF         = 0x0020,
+				ASSEMSPEC     = 0x0080,
+				ERROR         = 0x0100,
+			 }
+
+			ParseState m_parseState;
```

Comment:
ParseState does not seem to be used for anything. Delete it?

Reply:
How do you mean? `CheckParseState` checks against this value.

Reply:
`CheckParseState` checks this value, but the only thing that is done as a result of this test is setting this value to a different value.

Reply:
Maybe this should be debug-only and only used for asserts?

Reply:
Plan what to do with the `parseState` ? I would delete it completely. The asserts that it is used for are not very valuable.

Reply:
I want to keep it - The asserts helped me catch a porting typo in ToString, and they make some functions invariants explicit. I did remove the "error" state.

Reply:
Make it debug-only at least so that we do not have to carry it in shipping builds?

Reply:
Also, make the names follow the C# coding style.

Reply:
How do I make this debug-only? Is using System.Diagnostics.Debug not enough?

Reply:
There is still going to be the enum type, an empty method and 10+ calls to this method...